File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    double myPow(double x, int pow) {

        long long n= pow;

        bool neg=false;
        if(n<0){

            neg=true;
            n=n*-1;
        }

        double a=x,ans=1;
        while(n>0)
        {
            if(n&1==1)
            {
                ans*=a;
            }
            a*=a;
            n=n>>1;
        }

        if(neg)
        {
            return 1/ans;
        }else
        {
            return ans;
        }

    }
};

File: 0050-powx-n/0050-powx-n.java
class Solution {
    double func(double x,long n)
    {
        if(n==0) return 1;
        if(n%2==0)
        {
            double v=func(x,n/2);
            return v*v;
        }else
        {
            double v=func(x,(n-1)/2);

            return x*v*v ;
        }
    }
    public double myPow(double x, int n) {
        long num=Math.abs((long)n);
        double ans=func(x,num);

        if(n>=0) return ans;

        return (double)1/ans;
    }
}